File: backend/app/schemas/shopify.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ShopifyJobStatus = Literal["pending", "processing", "succeeded", "failed"]


class ShopifySyncRequest(BaseModel):
    action: Literal["sync", "unpublish"] = "sync"


class ShopifyBatchSyncRequest(BaseModel):
    product_ids: list[uuid.UUID] = Field(min_length=1, max_length=100)


class ShopifyJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_type: str
    product_id: uuid.UUID
    status: ShopifyJobStatus
    attempts: int
    max_attempts: int
    available_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    last_error: str | None
    idempotency_key: str


class ShopifyMappingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    product_id: uuid.UUID
    shopify_product_id: str | None
    shopify_handle: str | None
    sync_status: str
    last_sync_at: datetime | None
    last_sync_hash: str | None
    last_error: str | None


class SyncLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID
    shopify_product_id: str | None
    action: str
    attempt: int
    status: str
    error_code: str | None
    error_message: str | None
    started_at: datetime
    completed_at: datetime | None


class ShopifySyncStatusResponse(BaseModel):
    product_id: uuid.UUID
    mapping: ShopifyMappingResponse | None
    jobs: list[ShopifyJobResponse]
    logs: list[SyncLogResponse]

File: frontend/src/types/shopify.ts
export type ShopifyJobStatus = "pending" | "processing" | "succeeded" | "failed";

export type ShopifyJob = {
  id: string;
  job_type: string;
  product_id: string;
  status: ShopifyJobStatus;
  attempts: number;
  max_attempts: number;
  available_at: string;
  started_at: string | null;
  completed_at: string | null;
  last_error: string | null;
  idempotency_key: string;
};

export type ShopifyMapping = {
  product_id: string;
  shopify_product_id: string | null;
  shopify_handle: string | null;
  sync_status: string;
  last_sync_at: string | null;
  last_sync_hash: string | null;
  last_error: string | null;
};

export type SyncLog = {
  id: string;
  product_id: string;
  shopify_product_id: string | null;
  action: string;
  attempt: number;
  status: string;
  error_code: string | null;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
};

export type ShopifySyncStatus = {
  product_id: string;
  mapping: ShopifyMapping | null;
  jobs: ShopifyJob[];
  logs: SyncLog[];
};
